### ISSUE TITLE
Several terms are used for the value of record type: objects, maps, and records

### ISSUE THREAD

**Issue by tom-tan** [common-workflow-language/common-workflow-language#775]
In the spec, several terms are used to indicate the value of record types.
We have to use the same term for the value of record types.

For example,
- The [section 4.1 in the spec](https://www.commonwl.org/v1.0/CommandLineTool.html#Input_binding) says:
  > Where the input type is a record, array, or map, recursively walk the schema and input object, ...

  However, there are no description about `map`. I guess it is same as a value of `record` types.

- The [section 5.1.2 in the spec](https://www.commonwl.org/v1.0/CommandLineTool.html#CommandLineBinding) says:
  > **object**: Add `prefix` only, and recursively add object fields for which `inputBinding` is specified.

  The term `object` is defined in [the section 2.1](https://www.commonwl.org/v1.0/CommandLineTool.html#Data_concepts). However, the term `object` used in 5.1.2 indicate a value of `record` types rather than the data structure equivalent to JSON object.

**Comment by mr-c** [comment]
Hello @tom-tan! I agree that the spec should be more consistent in this regard

**Comment by tom-tan** [comment]
Here is a list of terms and their definitions in CWL v1.2 that I want to fix or unify:

- `object` (defined in [2.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Data_concepts))
  > An **object** is a data structure equivalent to the "object" type in JSON, consisting of an unordered set of name/value pairs (referred to here as **fields**) and where the name is a string and the value is a string, number, boolean, array, or object.
- `record` (not explicitly defined)
  - Used in [4.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Input_binding) to indicate the type of objects
  - related: [`CommandInputRecordSchema`](https://www.commonwl.org/v1.2/CommandLineTool.html#CommandInputRecordSchema) etc
- `map` (not explicitly defined)
  - Used in [4.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Input_binding) to indicate the type of objects
  - I guess it is same as `record`
  - related: [map notation](https://www.commonwl.org/v1.2/CommandLineTool.html#map)

---
- `field` (defined in [2.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Data_concepts))
  > ... an unordered set of name/value pairs (referred to here as **fields**) and where the name is a string and the value is a string, number, boolean, array, or object.
- `parameter` (defined in [3.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Execution_concepts))
  > A **parameter** is a named symbolic input or output of process, with an associated datatype or schema. During execution, values are assigned to parameters to make the input object or output object used for concrete process invocation. 
  - Note that there are similar but different term: `input parameter` and `output parameter` in [2.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Data_concepts)
    > The fields of the input object are referred to as "input parameters".
    > The fields of the output object are referred to as "output parameters".

    where the `field` is just a name/value pair as shown above.

**Comment by tom-tan** [comment]
- `input schema`, `output schema` (defined in [2.1](https://www.commonwl.org/v1.2/CommandLineTool.html#Data_concepts))
  > An **input schema** describes the valid format (required fields, data types) for an input object.
  > An **output schema** describes the valid format for an output object.

**Comment by tetron** [comment]
You are right that object/record/map are describing (almost) the same thing, "object" is the "O" in JSON, "record" and "map" come from Avro (the distinction is significant for binary serialization, but irrelevant for CWL).  I don't think we can use the same term everywhere, but we can certainly add clarification text.

**Comment by tom-tan** [comment]
Can I send a pull request to use `record` instead of `map`, and to clarify that the type of `object` is referred to as `record` type?
A `map` for the data type is only used in 4.1 and we can safely replace or remove to use `record`.

---
In this issue, I leave other terms such as `field`, `parameter` and `schema` to focus on fixing the term `object`, `record` and `map`.

**Comment by tetron** [comment]
Just to complicate things further, @GlassOfWhiskey is working on a PR to the Avro map type to schema salad:  https://github.com/common-workflow-language/schema_salad/pull/672

The distinction is:

* "object" is the underlying schema-less data model of string keys with arbitrary JSON values
* "record" is a specialization of the "object" data model but restricted by a schema that limits it to specific keys and type of values ("fields")
* "map" is a specialization of the "object" data model that does not restrict keys, but only restricts the type of values

Maybe that will be helpful to using the correct language everywhere.

**Comment by tom-tan** [comment]
@tetron Thank you for the information!

I will check the current usage of terms and re-define them if needed according to your information.
